clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocol.java

```diff
@@ -59,67 +69,156 @@
+    public static final String OWNED_PARTITIONS_KEY_NAME = "owned_partitions";
+    public static final String REVOKED_PARTITIONS_KEY_NAME = "revoked_partitions";
     public static final String TOPIC_PARTITIONS_KEY_NAME = "topic_partitions";
     public static final String USER_DATA_KEY_NAME = "user_data";
 
     public static final short CONSUMER_PROTOCOL_V0 = 0;
+    public static final short CONSUMER_PROTOCOL_V1 = 1;
+
     public static final Schema CONSUMER_PROTOCOL_HEADER_SCHEMA = new Schema(
             new Field(VERSION_KEY_NAME, Type.INT16));
     private static final Struct CONSUMER_PROTOCOL_HEADER_V0 = new Struct(CONSUMER_PROTOCOL_HEADER_SCHEMA)
             .set(VERSION_KEY_NAME, CONSUMER_PROTOCOL_V0);
+    private static final Struct CONSUMER_PROTOCOL_HEADER_V1 = new Struct(CONSUMER_PROTOCOL_HEADER_SCHEMA)
+            .set(VERSION_KEY_NAME, CONSUMER_PROTOCOL_V1);
+
+    public static final Schema TOPIC_ASSIGNMENT_V0 = new Schema(
+        new Field(TOPIC_KEY_NAME, Type.STRING),
+        new Field(PARTITIONS_KEY_NAME, new ArrayOf(Type.INT32)));
 
     public static final Schema SUBSCRIPTION_V0 = new Schema(
             new Field(TOPICS_KEY_NAME, new ArrayOf(Type.STRING)),
             new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES));
-    public static final Schema TOPIC_ASSIGNMENT_V0 = new Schema(
-            new Field(TOPIC_KEY_NAME, Type.STRING),
-            new Field(PARTITIONS_KEY_NAME, new ArrayOf(Type.INT32)));
+
+    public static final Schema SUBSCRIPTION_V1 = new Schema(
+        new Field(TOPICS_KEY_NAME, new ArrayOf(Type.STRING)),
+        new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES),
+        new Field(OWNED_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)));
+
     public static final Schema ASSIGNMENT_V0 = new Schema(
             new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)),
             new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES));
 
-    public static ByteBuffer serializeSubscription(PartitionAssignor.Subscription subscription) {
+    public static final Schema ASSIGNMENT_V1 = new Schema(
+        new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)),
+        new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES),
+        new Field(REVOKED_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)),
+        ERROR_CODE);
+
+    public enum Errors {
+        NONE(0),
+        NEED_REJOIN(1);
+
+        private final short code;
+
+        Errors(final int code) {
+            this.code = (short) code;
+        }
+
+        public short code() {
+            return code;
+        }
+
+        public static Errors fromCode(final short code) {
+            switch (code) {
+                case 0:
+                    return NONE;
+                case 1:
+                    return NEED_REJOIN;
+                default:
+                    throw new IllegalArgumentException("Unknown error code: " + code);
```

Comment:
Is throwing an error on unknown exception appropriate here? I can imagine future versions wanting to make use of this field by adding new error codes, and this will break the forwards compatibility in place elsewhere (ie that higher versions can be parsed as V1)

Reply:
Hmm interesting point. If we want to maintain forward compatibility we need to consider what's an appropriate default behavior for unknown error code.

On the other hand, as mentioned in https://cwiki.apache.org/confluence/display/KAFKA/KIP-429%3A+Kafka+Consumer+Incremental+Rebalance+Protocol#KIP-429:KafkaConsumerIncrementalRebalanceProtocol-LookingintotheFuture we'd likely change the protocol as future works so that broker will try to pick the leader with the highest version, and hence as long as the leader is "backward compatible" meaning it knows all other member's recognizable protocol and makes sure to now give them unknown information, this may not be a bad situation either. Thoughts?

Reply:
It seems a bit messy to require the leader to check each member's protocol against a list of acceptable error codes for that version. I think just logging on some level rather than throwing an exception provides a more flexible path forward without relying on future work, but I do see your point.

Reply:
If the newly added error code in the future requires consumer clients to take some action, then just logging an error and proceed by default may not be the best solution.

On the other hand, I can see that requiring leader to always keep track of other member's versions is indeed a bit intrusive (e.g. at the moment this version information is only used for serde at the `ConsumerProtocol`, and when ti was given to the Subscription / Assignment it was omitted already).

Regarding the code messiness, yeah that's also another thing to keep in mind, today we do not have a strict policy to "deprecate" older versions across all modules i.e. we let all newer versions to be backward compatible to all older versions, I think moving forward we can gradually remove the logic for handling older versions?

Given that, so how about this: I can augment `Subscription / Assignment` to always have the version information in this PR, and hence moving forward they will always be available to leader's assignor beyond version 2.3, and hence making such backward compatible logic would not be too hard.

Reply:
SGTM